sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs

```diff
@@ -0,0 +1,154 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class CallToAsyncMethodShouldNotBeBlocking : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4462";
+        private const string MessageFormat = "Replace this use of '{0}' with '{1}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly Dictionary<string, HashSet<KnownType>> InvalidMemberAccesses =
+            new Dictionary<string, HashSet<KnownType>>
+            {
+                ["Wait"] = new HashSet<KnownType> { KnownType.System_Threading_Tasks_Task },
+                ["WaitAny"] = new HashSet<KnownType> { KnownType.System_Threading_Tasks_Task },
+                ["WaitAll"] = new HashSet<KnownType> { KnownType.System_Threading_Tasks_Task },
+                ["Result"] = new HashSet<KnownType> { KnownType.System_Threading_Tasks_Task_T },
+                ["Sleep"] = new HashSet<KnownType> { KnownType.System_Threading_Thread },
+                ["GetResult"] = new HashSet<KnownType> { KnownType.System_Runtime_CompilerServices_TaskAwaiter,
+                    KnownType.System_Runtime_CompilerServices_TaskAwaiter_TResult },
+            };
+
+        private static readonly Dictionary<string, KnownType> DeadlockPreventingConstructs =
+            new Dictionary<string, KnownType>
+            {
+                ["StartNew"] = KnownType.System_Threading_Tasks_TaskFactory,
+                ["Run"] = KnownType.System_Threading_Tasks_Task,
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    if (c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var simpleMemberAccess = (MemberAccessExpressionSyntax)c.Node;
+                    var memberAccessNameName = simpleMemberAccess.Name?.Identifier.ValueText;
+
+                    if (memberAccessNameName == null ||
```

Comment:
The logic in this and the next if statement is difficult to follow. I think it could/should be simplified.